### PR TITLE
Warns when single line text fields overflow

### LINKED ISSUE #35706: Visual indication in debug mode when  Text in TextField isn't Center Align
I tried finding in a lot of resources but unfortunately i could not find a way to align the text vertically centre in a textfield. I also tried using suffixIcon instead of suffix but still not luck. Here is my code :

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _HomePageState();
  }
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: Icon(
          Icons.menu,
          color: Colors.black,
        ),
        backgroundColor: Colors.white,
        title: Container(
          margin: EdgeInsets.only(bottom: 10),
          child: Image.asset(
            "icons/logo.png",
          ),
        ),
        bottom: PreferredSize(
          child: Padding(
            padding: EdgeInsets.only(
              left: 10,
              right: 10,
              bottom: 10,
            ),
            child: Container(
              height: 40,
              child: TextField(
                textAlignVertical: TextAlignVertical.center,
                textAlign: TextAlign.left,
                maxLines: 1,
                style: TextStyle(
                  fontSize: 13,
                ),
                decoration: InputDecoration(
                    suffixIcon: IconButton(icon: Icon(Icons.search, color: Colors.black,), onPressed: (){}),
                    border: OutlineInputBorder(
                      borderSide: BorderSide(
                        color: Colors.black,
                      ),
                      borderRadius: BorderRadius.all(Radius.circular(15)),
                    )
                ),
              ),
            ),
          ),
          preferredSize: Size(MediaQuery.of(context).size.width, 50),
        ),
      ),
      body: Container(
        margin: EdgeInsets.only(top: 11),
        child: Column(
          children: <Widget>[
            Carousel(),
          ],
        ),
      ),
    );
  }
}

class Carousel extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _CarouselState();
  }
}

class _CarouselState extends State<Carousel> {
  List<String> urls = [];

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.symmetric(horizontal: 10),
      child: Stack(
        children: <Widget>[
          Image.network(
              "someImageUrlHere."),
          Positioned(
            bottom: 5,
            width: MediaQuery.of(context).size.width - 20,
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Text("•"),
                Text("•"),
                Text("•"),
                Text("•"),
                Text("•"),
              ],
            ),
          ),
        ],
      ),
    );
  }
}
```

What could be the issue that is causing this problem ? and how can i solve this issue ?

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

Warns if a single line text field's parent is not able to provide sufficient height to accommodate the text field. So developers don't get surprised when they see weird clipping. Set `clipBehavior` to `Clip.none` to opt-out (`TextField` currently doesn't seem to have a `clipBehavior` parameter).

## Related Issues

Alleviates https://github.com/flutter/flutter/issues/51257. Not sure that issue really needs a fix. It happens because we clip the `RenderParagraph` when the text gets too long along the scroll axis (horizontal for single line and vertical for multiline), but as a side effect the other axis gets clipped as well.
Fixes https://github.com/flutter/flutter/issues/35706

## Tests

I added the following tests:

- RenderEditable can stretch vertically

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 15 untriaged digest(s) on patchset 5.
View them at https://flutter-gold.skia.org/cl/github/63639

**Comment by skia-gold** [comment]
Gold has detected about 15 untriaged digest(s) on patchset 5.
View them at https://flutter-gold.skia.org/cl/github/63639

**Comment by skia-gold** [comment]
Gold has detected about 15 untriaged digest(s) on patchset 6.
View them at https://flutter-gold.skia.org/cl/github/63639

**Comment by skia-gold** [comment]
Gold has detected about 15 untriaged digest(s) on patchset 6.
View them at https://flutter-gold.skia.org/cl/github/63639

**Comment by skia-gold** [comment]
Gold has detected about 15 untriaged digest(s) on patchset 6.
View them at https://flutter-gold.skia.org/cl/github/63639

**Comment by skia-gold** [comment]
Gold has detected about 15 untriaged digest(s) on patchset 7.
View them at https://flutter-gold.skia.org/cl/github/63639

**Comment by LongCatIsLooong** [comment]
@justinmc @HansMuller I'd like to get a high-level review of this idea. Could there be cases where a single-line text field is intentionally constrained to a small height so the text gets cut off?

**Comment by justinmc** [comment]
I'm on board with this at a high level. I like the way you implemented it, maybe it'll be useful elsewhere in the future.  I can't think of a reason off the top of my head that you would want to purposely have a text field overflow, but interested to hear what Hans says.

**Comment by LongCatIsLooong** [comment]
I put this on frob: cl/326752193.

**Review by justinmc — COMMENTED** [review]
Just some small code changes and some questions, but overall this looks good. I think the approach of using composition and RenderConstraintsTransformBox is really solid and reusable.  We'll have to watch for this breaking anything else.

How did the tests look when you used frob by the way? (I can't figure out how to see the test results)

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 1]
```diff
@@ -741,6 +734,78 @@ class RenderUnconstrainedBox extends RenderAligningShiftedBox with DebugOverflow
...
+    }
+    switch (constrainedAxis) {
+      case Axis.horizontal:
+        return constraints.copyWith(maxHeight: double.infinity, minHeight: 0);
+      case Axis.vertical:
+        return constraints.copyWith(maxWidth: double.infinity, minWidth: 0);
+    }
+    assert(false);
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/rendering/shifted_box.dart:806]: Ah got it, that makes sense.

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 2]
```diff
@@ -741,6 +734,78 @@ class RenderUnconstrainedBox extends RenderAligningShiftedBox with DebugOverflow
   }
 }
 
+/// Renders a box, imposing no constraints on its child, allowing the child to
+/// render at its "natural" size.
+///
+/// This allows a child to render at the size it would render if it were alone
+/// on an infinite canvas with no constraints. This box will then attempt to
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/shifted_box.dart:741]: Alone on an Infinite Canvas with No Constraints will be the name of my breakout studio album.

**Review by justinmc — APPROVED** [review]
LGTM as long as Frob says it's not going to break any tests 👍

**Review by HansMuller — COMMENTED** [review]
This is a nice piece of work! Just some small-time feedback, mostly wordsmithing.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -2218,6 +2219,115 @@ class ConstrainedBox extends SingleChildRenderObjectWidget {
...
+  ///
+  ///  * [Alignment] for non-[Directionality]-aware alignments.
+  ///  * [AlignmentDirectional] for [Directionality]-aware alignments.
+  final AlignmentGeometry alignment;
+
+  /// @{template flutter.widgets.constraintsTransform}
+  /// The function used to transform the [BoxConstraints] imposed on [child].
+  ///
+  /// Must not be null. The function must not return null. The result will be
+  /// normalized.
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/basic.dart:2303]: Makes sense. Removing the `normalize()` call from `performLayout`.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -2218,6 +2219,115 @@ class ConstrainedBox extends SingleChildRenderObjectWidget {
...
+       super(key: key, child: child);
+
+  /// The text direction to use when interpreting the [alignment] if it is an
+  /// [AlignmentDirectional].
+  final TextDirection textDirection;
+
+  /// The alignment to use when laying out the child.
+  ///
+  /// If this is an [AlignmentDirectional], then [textDirection] must not be
+  /// null.
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/basic.dart:2291]: It does. Copied the docs from unconstrainedbox and didn't check.

**Comment by LongCatIsLooong** [comment]
updated

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**